src/rust/src/contract.rs

```diff
@@ -315,15 +358,27 @@ fn invoke_host_function_or_maybe_panic(
         Ok(rv) => xdr_to_rust_buf(&rv)?,
         Err(err) => {
             debug!(target: TX, "invocation failed: {}", err);
-            return Err(err.into());
+            return Ok(InvokeHostFunctionOutput {
+                success: false,
+                result_value: RustBuf { data: vec![] },
+                contract_events: vec![],
+                diagnostic_events: extract_diagnostic_events(&events)?,
+                modified_ledger_entries: vec![],
+                cpu_insns: 0,
+                mem_bytes: 0,
```

Comment:
Are we sure we want to return 0 here? 'Wasted' compute is a useful diagnostic on its own.

Reply:
Good question. I mentioned in the comment above the `InvokeHostFunctionOutput` definition that if `success` is false, the only field that will be populated is `diagnostic_events`. I agree that compute metrics would also be useful, but I wouldn't make that change until we decide to do something with that metric. I also haven't looked into if there are any issues with pulling info from `budget` on a failed call (this should be fine though).

Reply:
Yeah, that's ok to only propagate the events for now, but probably we could also write this depending on the presence of the diagnostic flag.